pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java

```diff
@@ -105,15 +108,69 @@ default void insertGroup(long questionId, PicklistGroupDef group, int order, lon
      * @return the option ids
      */
     default List<Long> insertOptions(long questionId, List<PicklistOptionDef> options, long revisionId) {
-        List<Long> optionIds = new ArrayList<>();
+        boolean nonTextTemplateFound = options.stream()
+                .anyMatch(o -> o.getTooltipTemplate() != null && o.getTooltipTemplate().getTemplateType() != TemplateType.TEXT);
+
+        if (nonTextTemplateFound) {
+            new DaoException("Only TEXT template type is supported for tooltips");
+        }
+
+        // Let's try to get all the template for all the options in a single batch
+        List<Template> templateList = Stream.of(
+                options.stream().map(o -> o.getOptionLabelTemplate()),
+                options.stream().map(o -> o.getTooltipTemplate()),
+                options.stream().map(o -> o.getDetailLabelTemplate()),
+                options.stream().map(o -> o.getNestedOptionsLabelTemplate())
+        ).flatMap(i -> i).collect(toList());
+
+        Long[] templateIds = getTemplateDao().insertTemplates(templateList, revisionId);
+        List<Long> templateIdList = Arrays.asList(templateIds);
+
+        JdbiPicklistOption jdbiOption = getJdbiPicklistOption();
+
+        // batch enter all the options
+        long[] optionIds = jdbiOption.insert(questionId,
+                options.stream().map(o -> o.getStableId()).iterator(),
+                templateIdList.listIterator(),
+                templateIdList.listIterator(options.size()),
+                templateIdList.listIterator(2 * options.size()),
+                options.stream().map(o -> o.isDetailsAllowed()).iterator(),
+                options.stream().map(o -> o.isExclusive()).iterator(),
+                Stream.iterate(0, i -> i + DISPLAY_ORDER_GAP).iterator(),
+                revisionId,
+                templateIdList.listIterator(3 * options.size()));
+
         int displayOrder = 0;
+        int j = 0;
         for (PicklistOptionDef option : options) {
             displayOrder += DISPLAY_ORDER_GAP;
-            long optionId = insertOption(questionId, option, displayOrder, revisionId);
-            optionIds.add(optionId);
+            option.setOptionId(optionIds[j++]);
+
+            insertNestedOptions(questionId, option, displayOrder, revisionId);
```

Comment:
We are not batch-entering the nested options. Could but figured diminishing returns.